File: app/routers/floor_traffic.py
# app/routers/floor_traffic.py

import logging
from datetime import date, datetime, timedelta
from fastapi import APIRouter, HTTPException, status, Query
from fastapi.encoders import jsonable_encoder
from postgrest.exceptions import APIError

from app.db import supabase
from app.models import (
    FloorTrafficCustomer,
    FloorTrafficCustomerCreate,
    FloorTrafficCustomerUpdate,
    MonthMetrics,
)

router = APIRouter()

def _safe_select(cols: list[str]) -> str:
    return "*"

async def _fetch_range(start: date, end: date):
    start_dt = datetime.combine(start, datetime.min.time())
    end_dt = datetime.combine(end, datetime.min.time()) + timedelta(days=1)
    try:
        res = (
            supabase
            .table("floor_traffic_customers")
            .select("*")
            .gte("visit_time", start_dt.isoformat())
            .lt("visit_time", end_dt.isoformat())
            .order("visit_time", desc=False)
            .execute()
        )
        data = res.data or []
        return data if isinstance(data, list) else []
    except APIError as e:
        logging.error("floor_traffic._fetch_range error: %s", e)
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Failed to fetch floor traffic data")

@router.get(
    "/today",
    response_model=list[FloorTrafficCustomer],
    summary="Get today's floor-traffic entries",
)
async def get_today_floor_traffic():
    today = date.today()
    return await _fetch_range(today, today)

@router.get(
    "/search",
    response_model=list[FloorTrafficCustomer],
    summary="Search floor-traffic by date range",
)
async def search_floor_traffic(
    start: date | None = Query(None, description="Start date (YYYY-MM-DD)"),
    end: date | None = Query(None, description="End date (YYYY-MM-DD)"),
):
    if start is None and end is None:
        start = end = date.today()
    elif start is None:
        start = end
    elif end is None:
        end = start

    return await _fetch_range(start, end)

@router.get(
    "/",
    response_model=list[FloorTrafficCustomer],
    summary="Alias for today's entries",
)
async def get_today_alias():
    return await get_today_floor_traffic()

@router.post(
    "/",
    response_model=FloorTrafficCustomer,
    status_code=status.HTTP_201_CREATED,
    summary="Log a new visitor",
)
async def create_floor_traffic(entry: FloorTrafficCustomerCreate):
    payload = jsonable_encoder(entry)

    if not payload.get("visit_time") or not payload.get("salesperson"):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="visit_time and salesperson are required",
        )

    first = payload.get("first_name")
    last = payload.get("last_name")
    if not first or not last:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="first_name and last_name are required",
        )
    payload["customer_name"] = f"{first.strip()} {last.strip()}"

    try:
        res = (
            supabase
            .table("floor_traffic_customers")
            .insert(payload)
            .execute()
        )
    except APIError as e:
        logging.error("create_floor_traffic insert error: %s", e)
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))

    if not res.data:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Database insertion failed, no data returned."
        )

    created = res.data[0]

    try:
        supabase.table("contacts").insert({
            "name": created.get("customer_name"),
            "email": created.get("email"),
            "phone": created.get("phone"),
        }).execute()
    except APIError:
        logging.warning("Failed to insert contact record; continuing without halting.")

    return created

@router.put("/{entry_id}", response_model=FloorTrafficCustomer)
async def update_floor_traffic(entry_id: int, entry: FloorTrafficCustomerUpdate):
    payload = {k: v for k, v in jsonable_encoder(entry).items() if v is not None}
    if not payload:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="No fields to update")
    try:
        res = (
            supabase.table("floor_traffic_customers")
            .update(payload)
            .eq("id", entry_id)
            .execute()
        )
    except APIError as e:
        logging.error("update_floor_traffic error: %s", e)
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    if not res.data:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Entry not found")
    return res.data[0]

@router.get(
    "/month-metrics",
    response_model=MonthMetrics,
    summary="Return month-to-date sales performance metrics.",
)
async def month_metrics():
    today = date.today()
    start = datetime.combine(today.replace(day=1), datetime.min.time())
    # Handle December → January edge-case
    if start.month == 12:
        end = start.replace(year=start.year + 1, month=1)
    else:
        end = start.replace(month=start.month + 1)

    try:
        res = (
            supabase
            .table("floor_traffic_customers")
            .select("*")
            .gte("visit_time", start.isoformat())
            .lt("visit_time", end.isoformat())
            .execute()
        )
        rows = res.data or []
    except APIError as e:
        logging.error("floor_traffic.month_metrics query failed: %s", e)
        return MonthMetrics(
            total_customers=0,
            demo_count=0,
            worksheet_count=0,
            customer_offer_count=0,
            sold_count=0
        )

    total = len(rows)
    demo = sum(1 for r in rows if r.get("demo"))
    worksheet = sum(1 for r in rows if any(r.get(k) for k in ["worksheet", "write_up", "worksheet_complete"]))
    customer_offer = sum(1 for r in rows if r.get("customer_offer"))
    sold = sum(1 for r in rows if r.get("sold"))

    return MonthMetrics(
        total_customers=total,
        demo_count=demo,
        worksheet_count=worksheet,
        customer_offer_count=customer_offer,
        sold_count=sold,
    )

File: routes/floorTraffic.js
// routes/floorTraffic.js
import express from 'express';
import FloorTrafficModel from '../models/FloorTrafficModel.js';
import supabase from '../lib/supabaseClient.js';

const router = express.Router();

/**
 * GET /api/floor-traffic/today
 * Fetch today’s floor-traffic entries.
 */
router.get('/floor-traffic/today', async (req, res, next) => {
  try {
    const todayLogs = await FloorTrafficModel.findToday();
    return res.json(todayLogs);
  } catch (err) {
    return next(err);
  }
});

/**
 * GET /api/floor-traffic
 * Alias for today's floor-traffic entries.
 * Mirrors the FastAPI implementation.
 */
router.get('/floor-traffic', async (req, res, next) => {
  try {
    const todayLogs = await FloorTrafficModel.findToday();
    return res.json(todayLogs);
  } catch (err) {
    return next(err);
  }
});

/**
 * GET /api/floor-traffic/month-metrics
 * Return month to date sales performance counts.
 */
router.get('/floor-traffic/month-metrics', async (req, res, next) => {
  try {
    const start = new Date();
    start.setDate(1);
    start.setHours(0, 0, 0, 0);
    const end = new Date(start);
    end.setMonth(end.getMonth() + 1);

    const { data, error } = await supabase
      .from('floor_traffic_customers')
      .select('*')
      .gte('visit_time', start.toISOString())
      .lt('visit_time', end.toISOString());
    if (error) throw error;

    const rows = data || [];
    const total = rows.length;
    const demo = rows.filter(r => r.demo).length;
    const writeUp = rows.filter(r =>
      r.worksheet || r.writeUp || r.worksheet_complete || r.worksheetComplete || r.write_up
    ).length;
    const sold = rows.filter(r => r.sold).length;

    res.json({
      totalCustomers: total,
      demoCount: demo,
      writeUpCount: writeUp,
      soldCount: sold,
    });
  } catch (err) {
    return next(err);
  }
});

/**
 * POST /api/floor-traffic
 * Create a new floor-traffic entry.
 */
router.post('/floor-traffic', async (req, res, next) => {
  try {
    const {
      timeIn,
      timeOut,
      salesperson,
      customerName,
      vehicle,
      trade,
      demo,
      writeUp,
      customerOffer,
      mgrTO,
      origin,
    } = req.body;

    // Basic validation
    if (!timeIn || !salesperson || !customerName) {
      return res
        .status(422)
        .json({ message: 'timeIn, salesperson & customerName are required' });
    }

    // Create the record (adjust for your ORM)
    const newLog = await FloorTrafficModel.create({
      timeIn,
      timeOut,
      salesperson,
      customerName,
      vehicle,
      trade,
      demo,
      writeUp,
      customerOffer,
      mgrTO,
      origin,
    });

    // Attempt to also create a contact record for this customer.
    // Ignore any errors so the floor-traffic entry still persists.
    const name = customerName || `${req.body.first_name ?? ''} ${req.body.last_name ?? ''}`.trim();
    if (name) {
      try {
        const { error } = await supabase
          .from('contacts')
          .insert({
            name,
            email: req.body.email || null,
            phone: req.body.phone || null,
          })
          .single();
        if (error) console.warn('Failed to insert contact record:', error.message);
      } catch (e) {
        console.warn('Failed to insert contact record:', e.message);
      }
    }

    return res.status(201).json(newLog);
  } catch (err) {
    return next(err);
  }
});

/**
 * PUT /api/floor-traffic/:id
 * Update an existing floor-traffic entry.
 */
router.put('/floor-traffic/:id', async (req, res, next) => {
  try {
    const updated = await FloorTrafficModel.update(req.params.id, req.body);
    if (!updated) return res.status(404).json({ message: 'Not Found' });
    return res.json(updated);
  } catch (err) {
    return next(err);
  }
});

export default router;
